js/create_progress_bar.js

```diff
@@ -0,0 +1,182 @@
+    }
+    if (!(parameters.container instanceof HTMLElement)) {
+        throw new Error(`Progress key must be an HTML element, "${parameters.container?.constructor?.name || typeof parameters.container}" found.`);
+    }
+
+    const main_container = document.createElement('div');
+    main_container.innerHTML = `
+        <div style="padding-left: 20px; padding-right: 20px;">
+            <div class="progress" style="height: 15px;">
+                <div class="progress-bar bg-info" role="progressbar" style="width:0;" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">0%</div>
+            </div>
+            <div class="messages_container"></div>
+        </div>
+    `;
+    const progress_bar = main_container.querySelector('.progress-bar');
+    const messages_container = main_container.querySelector('.messages_container');
+
+    parameters.container.appendChild(main_container);
+
+    const start_timeout = 250;
+
+    let is_running = true;
+    let abort_controller = new AbortController();
+
+    /**
+     * @param {null|number} percentage
+     */
+    function set_bar_percentage(percentage) {
+        progress_bar.style.width = `${typeof percentage === 'number' ? percentage : 0}%`;
+        progress_bar.innerHTML = typeof percentage === 'number' ? `${Math.floor(percentage)}%` : '-';
+        progress_bar.setAttribute('aria-valuenow', percentage || '0');
+
+    }
+
+    /**
+     * @param {number} value
+     * @param {number} max
+     * @param {null|string} text
+     */
+    function update_progress(value, max, text) {
+        value = value || 0;
+        max = max || 1;
+        const percentage = (value / max * 100);
+
+        set_bar_percentage(percentage);
+
+        if (text && text.length) {
+            messages_container.innerHTML = _.escape(text.trim()).replace(/\n/gi, '<br>');
+        }
+    }
+
+    function stop_progress_with_warning_state() {
+        is_running = false;
+
+        set_bar_percentage(null);
+        progress_bar.classList.remove('bg-info');
+        progress_bar.classList.add('bg-warning');
+    }
+
+    async function check_progress() {
+        is_running = true;
+        setTimeout(async () => {
+            try {
+                const res = await fetch(`/progress/check/${parameters.key}`, {
+                    method: 'POST',
+                    signal: abort_controller.signal,
+                });
+
+                if (res.status === 404) {
+                    const cb_err_result = parameters?.error_callback('Not found');
+                    if (!cb_err_result) {
+                        stop_progress_with_warning_state();
+                        return;
+                    }
+                }
+
+                if (res.status >= 300) {
+                    parameters?.error_callback(`Invalid response from server, expected 200 or 404, found "${res.status}".`);
+                    stop_progress_with_warning_state();
+                    return;
+                }
```

Comment:
I do not think it is necessary to distinguish status codes here. You may pass the status code to the error callback, but I am not even sure it is necessary.